---
subtitle: "TMA4268 Statistical Learning V2020"
title: "Module 8: Tree-based Mthods"
author: "Stefanie Muff, Department of Mathematical Sciences, NTNU"
date: "March xx, 2020"
fontsize: 10pt
output:
  beamer_presentation:
    keep_tex: yes
    fig_caption: false
    latex_engine: xelatex
    theme: "Singapore"
    colortheme: "default"
    font: "serif"
urlcolor: blue
bibliography: refs.bib
header-includes: \usepackage{multirow}

---

```{r setup, include=FALSE}
showsolA<-TRUE
showsolB<-TRUE
showsol<-TRUE
library(knitr)
opts_chunk$set(tidy.opts=list(width.cutoff=65),tidy=TRUE)
knitr::opts_chunk$set(echo = TRUE,tidy=TRUE,message=FALSE,warning=FALSE,strip.white=TRUE,prompt=FALSE,
                      cache=TRUE, size="scriptsize")
whichformat="latex"
```

# Introduction 

## Learning material for this module

* James et al (2013): An Introduction to Statistical Learning. Chapter 8.  
* Classnotes (todo: give link, or write on board)

<!-- * [Classnotes 04.03.2019](https://www.math.ntnu.no/emner/TMA4268/2019v/notes/M8L1notes.pdf)  -->
<!-- * [Classnotes 07.03.2019](https://www.math.ntnu.no/emner/TMA4268/2019v/notes/M8L2notes.pdf)  -->

Some of the figures in this presentation are taken (or are inspired) from @james.etal.

---

## What will you learn?

\vspace{2mm}

You will get to know

* Decision trees  
* Regression trees  
* Classification trees  
* Pruning a tree
* Bagging
* Variable importance
* Random forests
* Boosting

and learn how to apply all that.

--- 

## Example 1 (from chapter 8.1; `Hitters` data)
\vspace{1mm}

* Baseball players' salaries may depend on their experience (in years) and the number of hits.

* High salaries (yellow, red) vs low salaries (blue, green), salaries given on $\log$-scale. How can these be stratified for prediction of the salary? 

\centering
![Hits](hits.png){width=60%}

---

## Main idea of tree-based methods

* We can divide the area into rectangles with similar salaries. 

* Do this by deriving a set of decision (splitting) rules for segmenting the predictor space into a number of finer and finer regions. 


* All points in the same region will be given the same predictive value (the mean of all values in that square, or a majority vote).


 
In two dimensions (for two variables) a three-region partition may look like this:

\centering
![ISLR Figure 8.1](../../ISLR/Figures/Chapter8/8.2.png){width=50%}



---

The series of splitting rules can be visualized with a _regression tree_.

The following tree (which corresponds to the split in the previous slide) has three _leafs_ (terminal nodes), and two _internal nodes_:

$~$

\centering
 

![ISLR Figure 8.1](../../ISLR/Figures/Chapter8/8.1.png){width=50%}





--- 

## More than two predictors?

\vspace{2mm}

* With more than two predictors we cannot draw the partition of the data in a coordinate system, but we can still draw the regression tree. 

* Before we discuss how the algorithm splits the data into regions, let us look at a somewhat more interesting example.


---

## Example 2: Detection of Minor Head Injury
\tiny
(Artificial data)
\vspace{2mm}

\normalsize

* Data from patients that enter hospital. The aim is to quickly assess whether a patient as a brain injury or not.

* Patients are investigated and (possible) asked questions.

* Our job: To build a good model to predict quickly if someone has a brain injury. The method should be 

    +  **easy** to interpret for the medical personell that are not skilled in statistics, and 

    + **fast**, such that the medical personell quickly can identify a patient that needs treatment. 
    

    
$\rightarrow$ This can be done by using tree-based methods. 



\vspace{4mm}

\small
Note: Of course, the model should be built _before_ a new emergency patient arrives, using data that is already available.


---


The dataset includes data about 1321 patients and is a modified and smaller version of the (simulated) dataset `headInjury` from the `DAAG` library. 

\footnotesize

```{r,echo=FALSE}
library(DAAG)
options(digits=6)
headInjury2=read.table("headInjury2.txt",header=TRUE)
colnames(headInjury2)=c("amnesia","bskullf","GCSdecr","GCS.13","GCS.15","risk","consc","oskullf","vomit","brain.injury","age")

n=nrow(headInjury2)
set.seed(1)
train=sample(1:nrow(headInjury2),850)
test=setdiff(1:nrow(headInjury2),train)
headInjury2$brain.injury=factor(headInjury2$brain.injury)
for (i in 1:9) headInjury2[[i]]=as.factor(headInjury2[[i]])
#head(headInjury2[,c(1:9,11,10)])
head(headInjury2)
```
\normalsize

---


* The variable `brain.injury` will be the response of our model: It has value 1 if a person has an acute brain injury, and 0 otherwise. 

\vspace{0mm}

* 250 (19%) of the patients have a clinically important brain injury. 

\vspace{0mm}

* The 10 variables used as explanatory variables describe the state of the patient, for example

    + Is he/she vomiting?
    + Is the Glasgow Coma Scale (GCS) score\footnote{The GCS scale goes back to an article in the Lancet in 1974, and is used to describe the level of consciousness of patients with an acute brain injury. See <https://www.glasgowcomascale.org/what-is-gcs/>} after 2 hours equal to 15 (or not)?
    + Has he/she an open scull fracture?
    + Has he/she had a loss of consciousness?
    + and so on. 




---

The classification tree made from a training set of 850 randomly drawn observations (training set) for the head injury example looks like this:




```{r injury1, echo=FALSE, fig.width=7, fig.height=6,fig.align = "center",out.width='70%'}
library(tree)
headtree=tree(brain.injury~.,headInjury2,subset=train)
plot(headtree, type="uniform")
text(headtree,pretty=1)
```

<!-- ```{r injury2, echo=FALSE, fig.width=5, fig.height=4,fig.align = "center",out.width='70%'} -->
<!-- # competing method, gives different result due to other parameter defaults -->
<!-- library(rpart) -->
<!-- headrpart=rpart(brain.injury~.,headInjury2,subset=train,method="class") -->
<!-- plot(headrpart) -->
<!-- text(headrpart) -->
<!-- ``` -->

\small
Note: The split criterion at each node is to the left. For example, "GCS.15:0" means that "GCS.15=0" goes left, and "GCS.15=1" goes right.

---

\footnotesize
```{r,echo=TRUE,eval=TRUE}
print(headtree)
```
\normalsize

---

* By using simple decision rules related to the most important explanatory variables the medical staff can now assess the probability of a brain injury. 

* The decision can go "top down", because the most informative predictors are usually split first. 

* Example: The staff might check if the Glasgow Coma Scale of the patient is 15 after 2h, and if it was 13 at the beginning. In that case, the probability of brain injury is estimated to be 0.944 (node 7 in printout). 


**Advantages**:

* Decision trees are easier to interpret than many of the classification (and regression) methods that we have studied so far.

* Decision trees provide an easy way to visualize the data for non-statisticians.

---

## Glossary
* Classification and regression trees are usually drawn upside down, where the top node is called the *root*.

* The *terminal nodes* or *leaf nodes* are the nodes at the bottom, with no splitting criteria. These represent the final predicted class (for classification trees) or predicted response value (for regression trees) and are written symbolically as $R_j$ for $j = 1, 2, ..., J$ 

* The $R_j$ will be referred to as _non-overlapping regions_.

* *Internal nodes* are all nodes between the root and the terminal nodes. These nodes correspond to the partitions of the predictor space.

* *Branches*: segment of the tree connecting the nodes.

We will consider only binary splits on one variable, but multiway splits and linear combination of variabes are possible - but not so common.

---

# Constructing a decision tree 
You can construct decision trees for both classification and regression problems, first we focus on constructing a regression tree. 

## Regression tree (continous outcome)

Assume that we have a dataset consisting of $n$ pairs $(\boldsymbol{x}_i,y_i)$, $i=1,\ldots,n$, and each predictor is ${\boldsymbol{x}}_i=(x_{i1},x_{i2},...,x_{ip})$. The aim is to predict $y_i$.

\vspace{2mm}
Two steps:

1. Divide the predictor space into non-overlapping regions $R_1,R_2,\ldots,R_J$.

2. For every observation that falls into region $R_j$ we make the same prediction - which is the mean of the responses for the training observations that fall into $R_j$.

\vspace{2mm}

**But**: How to divide the predictor space into non-overlapping regions $R_1,R_2,\ldots,R_J$? 

---

We could try to minimize the RSS (residual sums of squares) on the training set given by

$$
\text{RSS}=\sum_{j=1}^J \sum_{i \in R_j}(y_i-\hat{y}_{R_j})^2,
$$

where $\hat{y}_{R_j}$ is the mean response for the training observations in region $j$. The mean $\hat{y}_{R_j}$ is also the predicted value for a new observations that falls into region $j$. 

To do this we need to consider every partition of the predictor space, and compute the RSS for each partition.

**But**: An exhaustive search over possible splits is _computationally infeasible_! In fact, constructing optimal binary decision tress is an NP-complete problem [@hyafilrivest1976].

---

<!-- Ripley (1996, p 216): Two types of optimality. a) Optimality of the partitioning of the predictor space : only feasible for small dimensions. b)  -->





## Recursive binary splitting 

\vspace{2mm}

* A _greedy_ approach is taken (aka top-down) - called _recursive binary splitting._ The idea is to find a split that minimizes RSS at each step.

* We start at the top of the tree and divide the predictor space into two regions, $R_1$ and $R_2$ by making a decision rule for one of the predictors $x_1, x_2,...,x_p$. If we define the two regions by $R_1(j,s)=\{x \mid x_j<s\}$ and $R_2(j,s)=\{x \mid x_j\geq s\}$, it means that we need to find the (predictor) $j$ and (splitting point) $s$ that minimize
$$\sum_{i: x_i \in R_1(j,s)}(y_i-\hat{y}_{R_1})^2+\sum_{i: x_i \in R_2(j,s)}(y_i -\hat{y}_{R_2})^2 \ ,$$
where $\hat{y}_{R_1}$ and $\hat{y}_{R_2}$ are the mean responses for the training observations in $R_1(j,s)$ and $R_2(j,s)$ respectively. This way we get the two first branches in our decision tree.

---

* We repeat the process to make branches further down in the tree. 

* For every iteration we let each single split depend on _only one of the predictors_, giving us two new branches. 

* This is done _successively_ and in each step we choose the split that gives the _\textcolor{red}{best split at that particular step}_, _i.e.,_ the split that gives the smallest RSS. 

* Continue splitting the predictor space until we reach some _stopping criterion_. For example we stop when a region contains less than 10 observations or when the reduction in the RSS is smaller than a specified limit.


\vspace{1mm}


**Q**: Why is this algorithm called _greedy_?

<!-- **A**: The winner takes it all, and don't consider splits that further down the tree might give a tree with a lower overall RSS. -->

---

## Regression tree: ozone example

\vspace{2mm}

Consider the `ozone` data set from the `ElemStatLearn` library. The data set consists of 111 observations on the following variables:

* `ozone` : the concentration of ozone in ppb
* `radiation`: the solar radiation (langleys)
* `temperature` : the daily maximum temperature in degrees F
* `wind` : wind speed in mph

```{r, echo=FALSE}
library(ElemStatLearn)
library(RColorBrewer)
myozone=ElemStatLearn::ozone
colnames(myozone)=c("ozone","radi","temp","wind")
```

```{r, echo=TRUE}
head(myozone)
```

---

Suppose we want to get an estimate of the ozone concentration based on the measurement of wind speed and the daily maximum temperature.



$\rightarrow$ We can fit a regression tree to the data, with `ozone` as our response variable and `temperature` and `wind` as predictors (not including radiation to make this easier to see). This gives us the following regression tree.

---

\footnotesize
```{r,echo=c(2:6,7)}
set.seed(300)
ozone.trainID = sample(1:111, 75)
ozone.train = myozone[ozone.trainID, ]
ozone.test = myozone[-ozone.trainID,]
# use the default settings in the tree function:
ozone.tree = tree(ozone~temp+wind, data=ozone.train)
```



```{r ozone1, echo=FALSE, fig.width=7, fig.height=6,fig.align = "center",out.width='60%'}
#par(pty="s")
plot(ozone.tree,type="uniform")
text(ozone.tree)
```


---

* We see that `temperature` is the "most important" predictor for predicting the ozone concentration. 
Observe that we can split on the same variable several times.

* Focus on the regions $R_j$, $j=1,\ldots, J$. What is $J$ here? Answer: $J=6$ (= the number of leaf nodes).

\vspace{2mm}
Below we see the partition of the `ozone` data set given in the `ozone.tree`, where 

* the ozone levels have been color-coded, where a darker color corresponds to a higher ozone concentration. 
* each rectangle corresponds to one leaf node. 
* each number corresponding to a leaf node has been found by taking an average of all observations (in the training set) in the corresponding region (rectangle). 
* each partition line corresponds to one internal node of a binary partition.



---

```{r, eval=TRUE,echo=FALSE}
# to make two figures next to eachother
par(mfrow=c(1,2),pty="s")
o.class = cut(ozone.train$ozone, breaks= 7)
col.oz = brewer.pal(9, "BuGn")
palette(col.oz[3:9])
plot(ozone.tree,type="uniform")
text(ozone.tree,cex=0.8)
plot(wind~temp, col=o.class, data=ozone.train, pch=20)
partition.tree(ozone.tree, add=TRUE,cex=0.8)
```


**Q**:

* Explain the connection between the tree and the region plot.
* Why is recursive binary splitting classified as a greedy algorithm?
* Discuss the advantages and disadvantages of letting each single split depend on only one of the predictors. 
* Does our tree automatically include interactions between variables?

---

**A:** 

* Each leaf node correspons to one region with prediction $\hat{y}_{R_j}$, and each internal note corresponds to a vertical or a horizontal line.
* Recursive splitting does not necessarily give the optimal global solution, but will give the best solution at each split (given what is done previously). 
* If the true connection between the response onzone and temperature and wind was so that splits should have been made on the diagonal of the temperature and wind space, that would take many splits on each of temperature and wind to produce. (See more below where we ask the same question for the classification setting.)
* Yes, we may have different effect of wind for different values of temperature.

---

## R: function `tree` in library `tree`

\vspace{2mm}

* R package `tree` by Brian D. Ripley [@tree2019].

* You get information about the function when you type `?tree` into the `R` console.

* Note: The default choice for a function to minimize is the deviance, and for normal data (as we may assume for regression), the deviance is proportional to the RSS.

<!-- --- -->

<!-- * Tree growth is limited to a depth of 31 by the use of integers to label nodes. -->

<!-- * Factor predictor variables can have up to 32 levels. This limit is imposed for ease of labelling, but since their use in a classification tree with three or more levels in a response involves a search over $2^{(k-1)} - 1$ groupings for $k$ levels, the practical limit is much less. -->

* A competing R function is `rpart`, explained in <https://cran.r-project.org/web/packages/rpart/vignettes/longintro.pdf>


---

### Stopping criterion
\vspace{2mm}

* When building the tree, we can (in principle) split until each leaf corresponds to one data point.
* Usually we use a less stringent stopping criterion, like the minimal number of nodes per region and/or the minimum reduction in the RSS.

* For example, the default in `tree()`: is given by `mincut=5, minsize=10, mindev=0.01`, thus a minimal size of 5 nodes and the minimal reduction in deviance (RSS) equal to 0.01. 
* We could make the ozone tree much deeper by decreasing the minimal size of a region and chanding the minimum reduction in RSS:

\scriptsize
```{r, results="asis"}
ozone.tree2 = tree(ozone~temp+wind, 
                  data=ozone.train, 
                  control = tree.control(75, mincut = 2, minsize = 4, mindev = 0.001)
                  )
```

---

With the above command, the tree would become very fine. Overfitting?

```{r ozone2, echo=FALSE, fig.width=7, fig.height=6,fig.align = "center",out.width='70%'}
plot(ozone.tree2)
text(ozone.tree2,cex=0.8)
```


---



## Tree performance

\vspace{2mm}

To test the predictive performance of our regression tree, we have randomly divided our observations into a test and a training set (here 1/3 test). For the two trees (the shallow and the deep one from above), we get the following test MSEs:

```{r}
ozone.pred = predict(ozone.tree, newdata=ozone.test)
ozone.MSE = mean((ozone.pred-ozone.test$ozone)^2)
ozone.MSE

ozone.pred2 = predict(ozone.tree2, newdata=ozone.test)
ozone.MSE2 = mean((ozone.pred2-ozone.test$ozone)^2)
ozone.MSE2
```

\vspace{1mm}

**How do we know if our tree performs good, or if there would be trees with a better predictive performance?**

---


# Pruning 

* If we have a data set with many predictors or choose a stringent stopping criterion, we may fit a (too) large tree. Thus, the number of observations from the training set that falls into some of the regions $R_j$ may be small $\rightarrow$ _\textcolor{red}{overfitting}_? 


* A smaller tree with fewer splits leads to fewer regions $R_1, \ldots, R_J$ with more observations. This might lead to lower variance and better interpretation at the cost of a little bias.

* One possible alternative to the process described above is to grow the tree only so long as the decrease in the RSS due to each split exceeds some (high) threshold.

* This strategy will result in smaller trees, but is too short-sighted: a seemingly worthless split early on in the tree might be followed by a very good split — that is, a split that leads to a large reduction in RSS later on.


---

## Cost complexity pruning 

* Better idea: to grow a very large tree $T_0$, and then _\textcolor{red}{prune}_ it back in order to obtain a subtree


* _\textcolor{red}{Cost complexity pruning}_ is used for this: We try to find a subtree $T\subset T_0$ that (for a given value of $\alpha$) minimizes
$$
C_{\alpha}(T)=Q(T)+\alpha |T|,
$$
where $Q(T)$ is our cost function, $|T|$ is the number of terminal nodes in tree $T$. The parameter $\alpha$ is then a parameter penalizing the number of terminal nodes, ensuring that the tree does not get too many branches. 


* For regression trees we choose 
$$Q(T)=\sum_{m=1}^{|T|}\sum_{x_i\in R_m}(y_i - \hat{y}_{R_m})^2 \ ,$$ 
and adapted cost functions for classification trees (the entropy (deviance), Gini or misclassification rate, see below).



---

* Given a value of $\alpha$ we get a pruned tree (note: the same pruned tree within a small range of $\alpha$).

* For $\alpha=0$ we get $T_0$ and as $\alpha$ increases we get smaller and smaller trees.


So, which value of $\alpha$ is best?

* We can use $K$-fold cross-validation to find out!

* Importantly, by increasing $\alpha$, branches get pruned in a _hierarchical (nested) fashion_.



Please study this [note from Bo Lindqvist in MA8701 in 2017 - Advanced topics in Statistical Learning and Inference](https://www.math.ntnu.no/emner/TMA4268/2018v/notes/CART1MA87012017BoLindqvist.pdf) for an example of how we perform cost complexity pruning in detail. Alterntively, this method, with proofs, are given in @Ripley, Section 7.2.


---

## Building a regression  tree: Algorithm 8.1

1. Use recursive binary splitting to grow a large tree on the training
data, stopping only when each terminal node has fewer than some
minimum number of observations.

2. Apply cost complexity pruning to the large tree in order to obtain a
sequence of best subtrees, as a function of $\alpha$.

3. Use $K$-fold cross-validation to choose $\alpha$. That is, divide the training
observations into $k$ folds. For each $k = 1,\ldots, K$:
      + Repeat Steps 1 and 2 on all but the $k$th fold of the training data.
      + Evaluate the mean squared prediction error on the data in the left-out $k$th fold, as a function of $\alpha$.
      + Average the results for each value of $\alpha$, and pick $\alpha$ to minimize the average error.
      
4. Return the subtree from Step 2 that corresponds to the chosen value of $\alpha$.


---

### Pruning the ozone tree

Let us start with the tree with many leafes, and then prune it with 5-fold CV. We can then plot the CV error as a function of tree size (instead of $\alpha$ -- why?):

```{r ozone.cv, echo=FALSE, fig.width=6, fig.height=5,fig.align = "center",out.width='60%'}
set.seed(563)
ozone.tree <- tree(ozone~temp+wind, 
                  data=ozone.train, 
                  control = tree.control(nrow(myozone), mincut = 2, minsize = 4, mindev = 0.001)
                  )
cv.ozone <- cv.tree(ozone.tree,K=5)
plot(cv.ozone$dev ~  cv.ozone$size,type= "b", lwd=2, col="red", xlab="Tree Size", ylab="Deviance")
```

Interestingly, it seems like a tree with 5 leafes performs best - which corresponds to the original choice.

<!-- --- -->

<!-- If we have a large dataset with many explanatory variables and terminal nodes, we can also prune the tree if we want to create a simpler tree and increase the interpretability of the model. -->

<!-- In the [classification tree](classtree2) (grown with deviance) we saw that we got several unnecessary splits. This is also something that can be avoided by pruning.  -->

<!-- In the [classification tree](classtree3) (grown with Gini index) there were 78 leafs - maybe we want a more easy interpretable tree? -->

<!-- But, there are many possible pruned versions of our full tree. Can we investigate all of these? -->



---

# Classification trees (binary or categorical outcome)

\vspace{2mm}

* Remember the minor head injury example: Binary outcome (disease yes/no).

\vspace{1mm}

* Now allow for $K\geq 2$ number of classes for the response.

\vspace{1mm}

* Building a decision tree in this setting is similar to building a regression tree for a quantitative response, but there are two main differences: _the prediction_ and _the splitting criterion_.

---

**1) The prediction:**   

* In the regression case we use the mean value of the responses in $R_j$ as a prediction for an observation that falls into region $R_j$. 

* For the _\textcolor{red}{classification case}_, however, we have two possibilities: 

    * Majority vote: Predict that the observation belongs to the most commonly occurring class of the training observations in $R_j$.  
    * Estimate the probability that an observation $x_i$ belongs to a class $k$, $\hat{p}_{jk}(x_i)$, and then classify according to a threshold value. This estimated probability is the proportion of class $k$ training observations in region $R_j$, with $n_{jk}$ observations. Region $j$ has $N_j$ observations. 
    $$\hat{p}_{jk} = \frac{1}{N_j} \sum_{i:x_i \in R_j} I(y_i = k)=\frac{n_{jk}}{N_j}.$$ 

---

**2) The splitting criterion:**  We do not use RSS as a splitting criterion for a qualitative variable. Instead we can use some _measure of impurity_ of the node. For leaf node $j$ and class $k=1,\ldots, K$:

* **Gini index**:
$$
G=\sum_{k=1}^K \hat{p}_{jk}(1-\hat{p}_{jk}) \ ,
$$
which is small if all of the $\hat{p}_{jk}$'s are close to 0 or 1.

\vspace{1mm}

* **Cross entropy**:
$$
D=-\sum_{k=1}^K \hat{p}_{jk}\log\hat{p}_{jk} \ .
$$
Here $\hat{p}_{jk}$ is the proportion of training observation in region $j$ that are from class $k$. 



When making a split in our classification tree, we want to minimize the Gini index or the cross-entropy.

---

Why don't we just minimize the misclassification error $$E= 1- \max_k{\hat{p}_{jk}} \ ?$$

**A**: 

* The Gini index and Entropy are _measure of impurity_. They are more sensitive to changes in the node probabilities. 

* Example for two classes: Assume we have two classes with 400 nodes each, written as $(400,400)$. Now we can choose between two splits:
    + Split 1: (100,300) and (300,100)  
    + Split 2: (200,400) and (200,0).

Both splits result in 25% misclassification. Which of these splits is better? Probably the second one, because it produces a pure node.



---

* Moreover, The Gini index and Entropy are differentiable (preferred for numerical optimization!)

```{r purity, echo=FALSE, fig.width=6, fig.height=5,fig.align = "center",out.width='60%'}
p=seq(0.001,0.999,length=1000)
gini=2*p*(1-p)
ent=-p*log(p)-(1-p)*log(1-p)
misc= pmin(1-p,p)
plot(p,misc,xlab="p",ylab="",ylim=c(0,0.5),type="l",main="K=2")
lines(p,ent*(0.5/max(ent)),col="red")
lines(p,gini,col="blue")
legend(0.3,0.2,legend=c("misclassification","cross-entropy","Gini"),fill=c("black","red","blue"))
```

---

### Minor head injury - split with cross-entropy

\vspace{1mm}

We use the 850 training samples to get a tree using cross-entropy (deviance):


\scriptsize
```{r}
tree.HIClass=tree(brain.injury~.,
                  data=headInjury2,
                  subset=train,split="deviance")
summary(tree.HIClass)

```

\normalsize

<!-- Deviance=$-2 \sum_{j}\sum_{k}n_{jk}\log(\hat{p}_{jk})$ for all nodes $j$ and classes $k$. -->
<!-- (Formula on page 255 of @MASS.) -->

\small
\vspace{1mm}
**Remark**: the deviance is a scaled version of the cross entropy: 
$$-2\sum_{k=1}^K n_{jk} \log\hat{p}_{jk}\ , \text{where} \quad   \hat{p}_{jk}=\frac{n_{jk}}{N_j}\ , $$ thus `split="deviance"` implies that we split according to the entropy criterion.

---


\scriptsize
```{r headInjury, echo=TRUE, fig.width=6, fig.height=5,fig.align = "center",out.width='50%'}
plot(tree.HIClass,type="proportional")
text(tree.HIClass,pretty=1)
```

\normalsize

* With `type="proportional" `, the length of branches are now proportional to the decrease in impurity.

* The classification tree has two terminal nodes with factor "0" originating from the same branch. Why do we get this "unnecessary" split? (Hint: Node purity)

---

### Minor head injury - split with Gini index

\vspace{1mm}
The same analysis with the Gini index:

\scriptsize
```{r}
tree.HIClassG=tree(brain.injury~.,headInjury2,
                  subset=train,split="gini")
summary(tree.HIClassG)
```



---

\scriptsize
```{r headInjury2, echo=TRUE, fig.width=8, fig.height=6,fig.align = "center",out.width='70%'}
plot(tree.HIClassG,type="proportional")
text(tree.HIClassG,pretty=1)
```

\normalsize

This is a very bushy tree!

---

### Checking predictions
\vspace{2mm}

We also use the classification tree to predict the status of the patients in the test set. 

\vspace{2mm}

With deviance: 
\tiny

```{r}
library(caret)
tree.pred=predict(tree.HIClass,headInjury2[test,],type="class")
confusionMatrix(tree.pred,reference=headInjury2[test,]$brain.injury)$table
```

\normalsize

With the Gini-index:

\tiny

```{r}
tree.predG=predict(tree.HIClassG,headInjury2[test,],type="class")
confusionMatrix(tree.predG,reference=headInjury2[test,]$brain.injury)$table
```

---

Study the confusion matrices on the previous slide:

* Which algorithm makes more errors?

* Is one type of mistakes more severe than the other? 

* Discuss if/how it is possible to change the algorithm in order to decrease the number of severe mistakes.



---

## Building a classification tree: Algorithm 8.1

\vspace{2mm}

* Like for regression trees, we would like to find classification trees that are good at prediction.

* Algorithm 8.1 can be used with misclassification, the Gini index, or cross-entropy instead of the RSS as quality measure.

---


### Finding an optimal classification tree 

We prune the classification tree that was built with the deviance criterion. We use the misclassification error to do so:

\scriptsize

```{r headInjury3, echo=TRUE, fig.width=5, fig.height=4,fig.align = "center",out.width='60%'}
set.seed(1)
cv.head=cv.tree(tree.HIClass, FUN= prune.misclass)
plot(cv.head$size,cv.head$dev,type="b",
                   xlab="Terminal nodes",ylab="Misclassifications")
```


\tiny
(Todo: Exercise with `cv.head=cv.tree(tree.HIClass,FUN=prune.tree)` , that is, pruning according to deviance; Note: I think it is not possible to prune according to the Gini index - why??)


---

The function `cv.tree` automatically does $10$-fold cross-validation. `dev` is here the number of misclassifications.

\scriptsize

```{r}
print(cv.head)
```

---

* We have done cross-validation on our training set of 850 observations. According to the plot, the number of misclassifications is as low for 7, 8 or 9 nodes, so we choose 7 terminal nodes as the smallest model (lowest variance). 

* We prune the classification tree according to this value:

\scriptsize
```{r}
prune.HIClass=prune.misclass(tree.HIClass,best=7) 
```

---

\scriptsize
```{r headInjury4, echo=TRUE, fig.width=6, fig.height=5,fig.align = "center",out.width='60%'}
plot(prune.HIClass)
text(prune.HIClass,pretty=1)
```


\normalsize
$\rightarrow$ No unnecessary splits left, and we have a simple and interpretable decision tree. 

How is the predictive performance of the model affected?

---

\scriptsize

```{r}
tree.pred.prune <- predict(prune.HIClass,headInjury2[test,],type="class")
confusionMatrix(tree.pred.prune,headInjury2[test,]$brain.injury)$table
```

\normalsize

We see that the misclassification rate is as small as before indicating that the pruned tree is as good as the original tree for the test data.

---

The same repeated for the Gini-grown tree - comment on what is done.

\scriptsize

```{r headInjury5, echo=TRUE, fig.width=6, fig.height=5,fig.align = "center",out.width='60%'}
set.seed(10)
cv.headG=cv.tree(tree.HIClassG,FUN=prune.misclass)
plot(cv.headG$size,cv.headG$dev,type="b",
                   xlab="Terminal nodes",ylab="Misclassifications")

```


---

<!-- Check which tree has the lowest misclassification: -->

<!-- \scriptsize -->

<!-- ```{r} -->
<!-- print(cv.headG) -->
<!-- ``` -->

\scriptsize

```{r headInjury6, echo=TRUE, fig.width=7, fig.height=5,fig.align = "center",out.width='70%'}
prune.HIClassG=prune.misclass(tree.HIClassG,best=7) 
plot(prune.HIClassG)
text(prune.HIClassG,pretty=1)
```


\normalsize

Compare this tree to the one from pruning the deviance-based tree.

<!-- --- -->

<!-- Misclassification rate -->

<!-- ```{r} -->
<!-- tree.pred.pruneG = predict(prune.HIClassG,headInjury2[test,],type="class") -->
<!-- confusionMatrix(tree.pred.pruneG,headInjury2[test,]$brain.injury)$table -->
<!-- ``` -->




---

### Questions: 

\vspace{2mm}

Discuss the bias-variance tradeoff of a regression tree when increasing/decreasing the number of terminal nodes, i.e: 

* What happens to the bias?
* What happens to the variance of a prediction if we reduce the tree size? 

<!-- --- -->

<!-- **A**: -->

<!-- As the tree size increase the bias will decrease, and the variance will increase. This is the same as any other method when we increase the model complexity. -->


---

# Trees versus linear models



* What if we have $x_1$ and $x_2$ and the true class boundary (two classes) is linear in $x_1$, $x_2$ space. How can we do that with our binary recursive splits?


* What about a rectangular boundary?


\centering
![Linear boundary ISL Figure 8.7](../../ISLR/Figures/Chapter8/8.7.png){width=50%}



---

# From trees to forests

**Advantages (+)**

* Trees automatically select variables.
* Tree-growing algorithms scale well to large $n$, growing a tree greedily.
* Trees can handle mixed features (continouos, categorical) seamlessly, and can deal with missing data.
* Small trees are easy to interpret and explain to people.
* Some believe that decision trees mirror human decision making.
* Trees can be displayed graphically.

**Disadvantages (-)**

* Large trees are not easy to interpret.
* Trees do not generally have good prediction performance (high variance).
* Trees are not very robust, a small change in the data may cause a large change in the final estimated tree

---

## What is next?

\vspace{2mm}


* **Bagging**: grow many trees (from bootstrapped data) and average - to get rid of the non-robustness and high variance by averaging.

* **Variable importance** - to see which variables make a difference (now that we have many trees).

* **Random forest**: inject more randomness (and even less variance) by just allowing a random selection of predictors to be used for the splits at each node.

* **Boosting**: make one tree, then another based on the residuals from the previous, repeat. The final predictor is a weighted sum of these trees.

---

But first, 

## Leo Breiman - the inventor of CART, bagging and random forests

Quotation from [Wikipedia](https://en.wikipedia.org/wiki/Leo_Breiman)

Leo Breiman (January 27, 1928 – July 5, 2005) was a distinguished statistician at the University of California, Berkeley. He was the recipient of numerous honors and awards, and was a member of the United States National Academy of Science.

Breiman's work helped to bridge the gap between statistics and computer science, particularly in the field of machine learning. His most important contributions were his work on classification and regression trees and ensembles of trees fit to bootstrap samples. Bootstrap aggregation was given the name bagging by Breiman. Another of Breiman's ensemble approaches is the random forest.

---

From [Breimans obituary](https://www.berkeley.edu/news/media/releases/2005/07/07_breiman.shtml)

BERKELEY – Leo Breiman, professor emeritus of statistics at the University of California, Berkeley.

"It is trite to say so, but Leo Breiman was indeed a Renaissance man, and we shall miss him greatly," said Peter Bickel, professor of statistics and chairman this summer of UC Berkeley's statistics department.

Breiman retired in 1993, but as a Professor in the Graduate School, he continued to get substantial National Science Foundation grants and supervised three Ph.D. students. Bickel said that some of Breiman's best work was done after retirement.

---

"In particular," said Bickel, "he developed one of the most successful state-of-the-art classification programs, 'Random Forest.' This method was based on a series of new ideas that he developed in papers during the last seven years, and it is extensively used in government and industry."

Breiman's best known work is considered to be "Classification and Regression Trees," a work in collaboration with three other scholars that facilitates practical applications, such as the diagnosis of diseases, from a multitude of symptoms.

---

#<a id="bag">Bagging</a>#

Decision trees often suffer from high variance. By this we mean that the trees are sensitive to small changes in the predictors: If we change the observation set, we may get a very different tree. 

Let's draw a new training set for our data and see what happens if we fit our full classification tree (deviance grown).

---

```{r,echo=showsol}
set.seed(33)
N=dim(headInjury2)[1]
train2=sample(1:N,850) #We draw a new training sample. The number of observations is the same as before.
tree.HIClass2=tree(brain.injury~.,
                   data=headInjury2,subset=train2) 
par(mfrow=c(1,2))
#short=c("amnesia","bskullf","GCSdecr","GCS.13","GCS.15","risk","consc","oskullf","vomit","brain","age")
plot(tree.HIClass);text(tree.HIClass,pretty=0)
plot(tree.HIClass2);text(tree.HIClass2,pretty=0)

```

---

This classification tree is constructed by using $850$ observations, just like the tree in the [classification trees section](classtree2), but we get two different trees that will give different predictions for a test set.

To reduce the variance of decision trees we can apply _bootstrap aggregating_ (_bagging_), invented by Leo Breiman in 1996 (see [references](ref)).

---

## Independent data sets

Assume we have $B$ i.i.d. observations of a random variable $X$ each with the same mean and with variance $\sigma^2$. We calculate the mean $\bar{X} = \frac{1}{B} \sum_{b=1}^B X_b$. The variance of the mean is
$$\text{Var}(\bar{X}) = \text{Var}\Big(\frac{1}{B}\sum_{b=1}^B X_b \Big) = \frac{1}{B^2} \sum_{b=1}^B \text{Var}(X_b) = \frac{\sigma^2}{B}.$$
By averaging we get reduced variance. This is the basic idea!

But, we will not draw random variables - we want to fit decision trees: $\hat{f}_1({\bf x}),\hat{f}_2({\bf x}),\ldots, \hat{f}_B({\bf x})$ and average those.
$$ \hat{f}_{avg}({\bf x})=\frac{1}{B}\sum_{b=1}^B \hat{f}_b({\bf x})$$

However, we do not have many independent data set - so we use _bootstrapping_ to construct $B$ data sets.

---

## Bootstrapping (from Module 5)

Problem: we want to draw samples from a population with distribution $f$.

But: we do not know $f$ and do not have a population to draw from, we only have our one sample.

Solution: we may use our sample as an empirical estimate for the distribution $f$ - by assuming that each sample point has probability $1/n$ for being drawn. 

Therefore: we draw with replacement $n$ observations from our sample - and that is our first _bootstrap sample_.

We repeat this $B$ times and get $B$ bootstrap samples - that we use as our $B$ data sets.

---

## Bootstrap samples and trees

For each bootstrap sample we construct a decision tree, $\hat{f}^{*b}(x)$ with $b=1,...,B$, and we then use information from all of the trees to draw inference. 

For a regression tree, we take the average of all of the predictions and use this as the final result:
$$
\hat{f}_{bag}(x)=\frac{1}{B}\sum_{b=1}^B \hat{f}^{*b}(x).
$$
For a classification tree we record the predicted class (for a given observation $x$) for each of the $B$ trees and use the most occurring classification (majority vote) as the final prediction - or alternatively average posterior probabilities for each class.

---

Originally, Breiman (1996) suggested to prune each tree, but later research has found that it is better to leave the trees at maximal size (a bushy tree), to make the trees as different from each other as possible.

The number $B$ is chosen to be as large as "necessary". An increase in $B$ will not lead to overfitting, and $B$ is not regarded as a tuning parameter. If a goodness of fit measure is plotted as a function of $B$ (soon) we see that (given that $B$ is large enough) increasing $B$ will not change the goodness of fit measure.

But first, a smart way to avoid doing cross-validation.

---

##<a id="oob">Out-of-bag error estimation</a>##

* We use a subset of the observations in each bootstrap sample. From Module 5 we know that the probability that an observation is in the bootstrap sample is approximately $1-e^{-1}$=`r 1-exp(-1)` (approximately 2/3).
* when an observation is left out of the bootstrap sample it is not used to build the tree, and we can use this observation as a part of a "test set" to measure the predictive performance and error of the fitted model, $f^{*b}(x)$. 

In other words: Since each observation $i$ has a probability of approximately 2/3 to be in a bootstrap sample, and we make $B$ bootstrap samples, then observation $i$ will be outside the bootstrap sample in approximately $B/3$ of the fitted trees. 

The observations left out are referred to as the _out-of-bag_ observations, and the measured error of the $B/3$ predictions is called the _out-of-bag error_. 


<!-- Based on these $B/3$ predictions we can make one single prediction (_OOB prediction_) and measure the prediction error (_OOB error_).  -->

---

##<a id="bagex">Example</a>##

We can do bagging by using the function _randomForest()_ in the _randomForest_ library.

\footnotesize

```{r}
library(randomForest)
set.seed(1)
bag=randomForest(brain.injury~.,
                 data=headInjury2,subset=train,
                 mtry=10,ntree=500,importance=TRUE)
bag$confusion
1-sum(diag(bag$confusion))/sum(bag$confusion[1:2,1:2])
```
\normalsize

The variable _mtry=10_ because we want to consider all $10$ predictors in each split of the tree. The variable $ntree=500$ because we want to average over $500$ trees.

---

Predictive performance of the bagged tree on unseen test data:
```{r}
yhat.bag=predict(bag,newdata=headInjury2[test,])
misclass.bag=table(yhat.bag,headInjury2[test,]$brain.injury)
print(misclass.bag)
1-sum(diag(misclass.bag))/(sum(misclass.bag))
```
We note that the misclassification rate has increased slightly for the bagged tree (as compared to our previous full and pruned tree). **In other examples an improvement is very often seen.**

---

## Prediction by majority vote vs. by averaging the probabilities 

Consider the case when you have grown $B$ classification tree with a binary response with classes 0 and 1. You might wonder which approach to choose to make a final prediction: majority vote or an average of the probabilities? Or would the prediction be the same in each case? 

The difference between these two procedures can be compared to the difference between the mean value and median of a set of numbers. If we average the probabilities and make a classification thereafter, we have the mean value. If we sort all of our classifications, so that the classifications corresponding to one class would be lined up after each other, followed by the classifications corresponding to the other class we obtain the median value. 

---

We examine this by an example: 

Suppose we have $B=5$ (no, $B$ should be higher - this is only for illustration) classification tree and have obtained the following 5 estimated probabilities: \{0.4, 0.4, 0.4, 0.4, 0.9 \}. If we average the probabilities, we get 0.5, and if we use a cut-off value of 0.5, our predicted class is 1. However, if we take a majority vote, using the same cut of value, the predicted classes will be \{0, 0, 0, 0, 1 \}. The predicted class, based on a majority vote, would accordingly be 0.

The two procedures thus have their pros and cons: By averaging the predictions no information is lost. We do not only get the  final classification, but the probability for belonging to the class 0 or 1. However, this method is not robust to outliers. By taking a majority vote, outliers have a smaller influence on the result. 

---

## When should we use bagging?

Bagging can be used for predictors (regression and classification) that are not trees, and according to Breiman (1996) 

* the vital element is the instability of the prediction method
* if perturbing the learning set can cause significant changes in the predictor constructed, then bagging can improve accuracy.

Breiman (1996) suggests that these methods should be suitable for bagging:

* neural nets, classification and regression trees, subset selection in linear regression

however not nearest neigbours - since 

* the stability of nearest neighbour classification methods with respect to perturbations of the data distinguishes them from competitors such as trees and neural nets.

---


## Variable importance plots

Bagging is an example of an _ensemble method_, so is boosting and random forests (to come next). For all of these methods many trees are grown and combined, and the predictive power can be highly improved. However, this comes at a cost of interpretability. Instead of having one tree, the resulting model consists of $B$ trees, where $B$ often is 300 or 500 (or maybe even 5000 when boosting). 

Variable importance plots show _the relative importance of the predictors:_ the predictors are sorted according to their importance, such that the top variables have a higher importance than the bottom variables. 
There are in general two types of variable importance plots: 

* variable importance based on decrease in node impurity and 
* variable importance based on randomization.

---

### Variable importance based on node impurity

The term *important* relates to _total decrease in the node impurity, over splits for a predictor_, and is defined differently for regression trees and classification trees.

**Regression trees:** 

* The importance of each predictor is calculated using the RSS. 
* The algorithm records the total amount that the RSS is decreased due to splits for each predictor (there may be many spits for one predictor for each tree). 
* This decrease in RSS is then averaged over the $B$ trees. The higher the decrease, the more important the predictor.

---

**Classification trees:** 

* The importance of each predictor is calculated using the Gini index. 
* The importance is the mean decrease (over all $B$ trees) in the Gini index by splits of a predictor.

R: `varImpPlot` (or `importance`) in `randomForest` with `type=2`.

---

**Auto data example**

```{r}
set.seed(4268)
data(mtcars)
mtcars.rf <- randomForest(mpg ~ ., data=mtcars, ntree=1000, 
                          keep.forest=FALSE,
                          importance=TRUE)
```

---

```{r}
varImpPlot(mtcars.rf,type=2,pch=20)
```

---

### Variable importance based on randomization

Variable importance based on randomization is calculated using the OOB sample. 

* Computations are carried out for one bootstrap sample at a time. 
* Each time a tree is grown the OOB sample is used to test the predictive power of the tree. 
* Then for one predictor at a time, repeat the following: 
     + permute the OOB observations for the $j$th variable $x_j$ and calculate the new OOB error. 
     + If $x_j$ is important, permuting its observations will decrease the predictive performance. 
* The difference between the two is averaged over all trees (and normalized by the standard deviation of the differences). 

R: `varImpPlot` (or `importance`) in `randomForest` with `type=1`.

---

```{r}
varImpPlot(mtcars.rf,type=1,pch=20)
```


---

The two types together: do they agree?

```{r}
varImpPlot(mtcars.rf,pch=20)
```

---

#<a id="rf">Random Forest</a>

If there is a strong predictor in the dataset, the decision trees produced by each of the bootstrap samples in the bagging algorithm becomes very similar: Most of the trees will use the same strong predictor in the top split. 

We have seen this for our example trees for the minor head injury example, the predictor _GCS.15.2hours_ was chosen in the top split every time. This is probably the case for a large amount of the bagged trees as well.

This is not optimal, because we get $B$ trees that are highly correlated. We don't get a large reduction in variance by averaging $\hat{f}^{*b}(x)$ when the correlation between the trees is high. In the [previous section](bagex) we actually saw a (marginal) decrease in the predictive performance for the bagged tree compared to the [pruned tree](expruning) and the [full tree](classtree2).

_Random forests_ is a solution to this problem and a method for decorrelating the trees. 

---

## The effect of correlation on the variance of the mean

The variance of the average of $B$ observations of i.i.d random variables $X$, each with variance $\sigma^2$ is $\frac{\sigma^2}{B}$. Now, suppose we have $B$ observations of a random variable $X$ which are identically distributed, each with mean $\mu$ and variance $\sigma^2$, but not independent. 

That is, suppose the variables have a positive correlation $\rho$
$$\text{Cov}(X_i, X_j) = \rho \sigma^2, \quad i \neq j.$$

---

The variance of the average is 
$$\begin{aligned} \text{Var}(\bar{X}) &= \text{Var}\Big( \frac{1}{B}\sum_{i=1}^B X_i \Big) \\
&= \sum_{i=1}^B \frac{1}{B^2} \text{Var} (X_i) + 2 \sum_{i=2}^B \sum_{j=1}^{i-1} \frac{1}{B} \frac{1}{B} \text{Cov} (X_i, X_j) \\
&= \frac{1}{B} \sigma^2 + 2 \frac{B(B-1)}{2}\frac{1}{B^2} \rho \sigma^2 \\
&= \frac{1}{B} \sigma^2 + \rho \sigma^2 - \frac{1}{B} \rho \sigma^2 \\
&= \rho \sigma^2 + \frac{1-\rho}{B}\sigma^2\\
&= \frac{1-(1-B)\rho}{B} \sigma^2\\ \end{aligned}$$

Check: $\rho=0$ and $\rho=1$? (Most negative values of $\rho$ will not give a positive definite covariance matrix. The covariance matrix is positive definite if $\rho>-1/(B-1)$.)

```{r,eval=FALSE,echo=FALSE}
varmean=function(B=100,rho,sigma2=1) return(sigma2*(1-(1-B)*rho)/B)
rhovec=seq(-1,1,length=100)
plot(rhovec,varmean(rho=rhovec),type="b")
```

---

The idea behind random forests is to _improve the variance reduction of bagging_ by reducing the correlation between the trees. 

The procedure is thus as in bagging, but with the important difference, that 

* at each split we are only allowed to consider $m<p$ of the predictors. 

A new sample of $m$ predictors is taken at each split and 

* typically $m\approx \sqrt p$ (classificaton) and $m=p/3$ (regression)

The general idea is at for very correlated predictors $m$ is chosen to be small.

---

The number of trees, $B$, is not a tuning parameter, and the best is to choose it large enough. 

If $B$ is sufficiently large (three times the number needed for the random forest to stabilize), the OOB error estimate is equvalent to LOOCV (Efron and Hastie, 2016, p 330).

---

![ISLR Figure 8.10, gene expression data set with 15 classes and 500 predictors](../../ISLR/Figures/Chapter8/8.10.png)

---

##<a id="exrf">Example</a>##

We decorrelate the trees by using the _randomForest()_ function again, but this time we set _mtry=3_. This means that the algorithm only considers three of the predictors in each split. We choose $3$ because we have $10$ predictors in total and $\sqrt{10}\approx 3$. 

```{r}
set.seed(1)

rf=randomForest(brain.injury~.,
                data=headInjury2,subset=train,
                mtry=3,ntree=500,importance=TRUE)

```

---

We check the predictive performance as before:
\footnotesize
```{r}
rf$confusion
1-sum(diag(rf$confusion[1:2,1:2]))/(sum(rf$confusion[1:2,1:2]))

yhat.rf=predict(rf,newdata=headInjury2[test,])

misclass.rf=table(yhat.rf,headInjury2[test,]$brain.injury)
print(misclass.rf)

1-sum(diag(misclass.rf))/(sum(misclass.rf))
```
\normalsize

The misclassification rate is slightly decreased compared to the [bagged tree](bagex) (and to the [pruned tree](expruning)).

---

By using the _varImpPlot()_ function we can study the importance of each predictor.

```{r}
varImpPlot(rf,pch=20)
varImpPlot(rf,pch=20,type=1)
varImpPlot(rf,pch=20,type=2)

```

As expected _GCS.15.2hours_ is a strong predictor along with _basal.skull.facture_ and _age_. This means that most of the trees will have these predictors in the top split.

---

**Iris example**

Variable importance plot for bagging:

```{r,echo=showsol,eval=TRUE}
# borrowed from https://gist.github.com/ramhiser/6dec3067f087627a7a85
library(dplyr)
library(ggplot2)
#library(ggpubr)
rf_out <- randomForest(Species ~ ., data=iris, mtry=4)
rf_out$confusion

# Extracts variable importance (Mean Decrease in Gini Index)
# Sorts by variable importance and relevels factors to match ordering
var_importance <- data_frame(variable=setdiff(colnames(iris), "Species"),
                             importance=as.vector(importance(rf_out)))
var_importance <- arrange(var_importance, desc(importance))
var_importance$variable <- factor(var_importance$variable, levels=var_importance$variable)

p <- ggplot(var_importance, aes(x=variable, weight=importance, fill=variable))
p <- p + geom_bar() + ggtitle("Variable Importance from Random Forest Fit")
p <- p + xlab("Demographic Attribute") + ylab("Variable Importance (Mean Decrease in Gini Index)")
p <- p + scale_fill_discrete(name="Variable Name")
p=p + theme(axis.text.x=element_blank(),
          axis.text.y=element_text(size=12),
          axis.title=element_text(size=16),
          plot.title=element_text(size=18),
          legend.title=element_text(size=16),
          legend.text=element_text(size=12))
p
```

---


**Iris example**

Variable importance plot for random forest:

```{r,echo=showsol,eval=TRUE}
# borrowed from https://gist.github.com/ramhiser/6dec3067f087627a7a85

rf_out=randomForest(Species~.,data=iris)
rf_out$confusion
var_importance <- data_frame(variable=setdiff(colnames(iris), "Species"),
                             importance=as.vector(importance(rf_out)))
var_importance <- arrange(var_importance, desc(importance))
var_importance$variable <- factor(var_importance$variable, levels=var_importance$variable)

pp <- ggplot(var_importance, aes(x=variable, weight=importance, fill=variable))
pp <- pp + geom_bar() + ggtitle("Variable Importance from Random Forest Fit")
pp <- pp + xlab("Demographic Attribute") + ylab("Variable Importance (Mean Decrease in Gini Index)")
pp <- pp + scale_fill_discrete(name="Variable Name")
pp=pp + theme(axis.text.x=element_blank(),
          axis.text.y=element_text(size=12),
          axis.title=element_text(size=16),
          plot.title=element_text(size=18),
          legend.title=element_text(size=16),
          legend.text=element_text(size=12))
pp
```

---

#<a id="boost">Boosting</a>

_Boosting_ is an alternative approach for improving the predictions resulting from a decision tree. 
We will only consider the description of boosting regression trees (and not classification trees) in this course.

In boosting the trees are grown _sequentially_ so that each tree is grown using information from the previous tree. 

* First build a decision tree with $d$ splits (and $d+1$ terminal notes).
* Next, improve the model in areas where the model didn't perform well. This is done by fitting a decision tree to the _residuals of the model_. This procedure is called _learning slowly_. 
* The first decision tree is then updated based on the residual tree, but with a weight. 
* The procedure is repeated until some stopping criterion is reached. Each of the trees can be very small, with just a few terminal nodes (or just one split).

---

**Algorithm 8.2: Boosting for regression trees**

1. Set $\hat{f}(x) = 0$ and $r_i = y_i$ for all $i$ in the training set.
2. For $b=1,2,...,B$, repeat:  
    a) Fit a tree $\hat{f}^b$ with $d$ splits ($d+1$ terminal nodes) to the training data.  
    b) Update $\hat{f}$ by adding in a shrunken version of the new tree: $$\hat{f}(x) \leftarrow \hat{f}(x)+\lambda \hat{f}^b(x).$$  
    c) Update the residuals, $$r_i \leftarrow r_i - \lambda \hat{f}^b(x_i).$$  
3. The boosted model is $\hat{f}(x) = \sum_{b=1}^B \lambda \hat{f}^b(x).$

Boosting has three tuning parameters which need to set, and can be found using cross-validation.

---

**Tuning parameters**

* The number of trees to be grown, $B$. The value of $B$ could be chosen using cross-validation. A too small value of $B$ would imply that much information is unused (remember that boosting is a slow learner), whereas a too large value of $B$ may lead to overfitting. 
* $\lambda$: This is a shrinkage parameter and controls the rate at which boosting learns. The role of $\lambda$ is to scale the new information, when added to the existing tree. We add information from the $b$-th tree to our existing tree $\hat{f}$, but scaled by the $\lambda$. Choosing a small value for $\lambda$ ensures that the algorithm learns slowly, but will require a larger tree ensemble. Typical values of $\lambda$ is 0.1 or 0.01.
* Interaction depth $d$: The number of splits in each tree. This parameter controls the complexity of the boosted tree ensemble (the level of interaction between variables that we may estimate). By choosing $d=1$ a tree stump will be fitted at each step and this gives an additive model.

---

#<a id="boostex">Example: Boston data set</a>

First - to get to know the data set we run through trees, bagging and random forests - before arriving at boosting. See also the ISLR book, Section 8.3.4.

### Data
```{r}
library(MASS)
library(tree)
set.seed(1)
train = sample(1:nrow(Boston), nrow(Boston)/2)
colnames(Boston)
head(Boston)
```

---

### Regression tree
```{r}
tree.boston=tree(medv~.,Boston,subset=train)
summary(tree.boston); plot(tree.boston)
text(tree.boston,pretty=0)
```

---

### Need to prune?

```{r}
cv.boston=cv.tree(tree.boston)
plot(cv.boston$size,cv.boston$dev,type='b')
```

Most complex tree selected.

---

### Pruning

Just to show pruning (even if most complex tree was selected).

```{r}
prune.boston=prune.tree(tree.boston,best=5)
plot(prune.boston)
text(prune.boston,pretty=0)
```

---

### Test error for full tree

```{r}
yhat=predict(tree.boston,newdata=Boston[-train,])
boston.test=Boston[-train,"medv"]
plot(yhat,boston.test, pch=20)
abline(0,1)
mean((yhat-boston.test)^2)
```

---

### Bagging

```{r}

library(randomForest)
set.seed(1)
bag.boston=randomForest(medv~.,data=Boston,subset=train,mtry=13,importance=TRUE)
bag.boston
```

Plotting predicted test values vs true values.

```{r}
yhat.bag = predict(bag.boston,newdata=Boston[-train,])
plot(yhat.bag, boston.test,pch=20)
abline(0,1)
mean((yhat.bag-boston.test)^2)
```

Error rate on test set for bagging

```{r,eval=FALSE,echo=FALSE}
bag.boston=randomForest(medv~.,data=Boston,subset=train,mtry=13,ntree=25)
yhat.bag = predict(bag.boston,newdata=Boston[-train,])
mean((yhat.bag-boston.test)^2)
```

---

### Random forest

```{r}
set.seed(1)
rf.boston=randomForest(medv~.,data=Boston,subset=train,mtry=6,importance=TRUE)
yhat.rf = predict(rf.boston,newdata=Boston[-train,])
mean((yhat.rf-boston.test)^2)
importance(rf.boston)
varImpPlot(rf.boston)
```

---

### Finally, boosting Boston

```{r}
library(gbm)
set.seed(1)
boost.boston=gbm(medv~.,data=Boston[train,],
                 distribution="gaussian",
                 n.trees=5000,interaction.depth=4)
summary(boost.boston,plotit=FALSE)
```
---

**Partial dependency plots - integrating out other variables **

```{r}
par(mfrow=c(1,2))
plot(boost.boston,i="rm")
plot(boost.boston,i="lstat")
```

---

**Prediction on test set**

First for model with $\lambda=0.001$ (default), then with $\lambda=0.2$: MSE on test set. We could have done cross-validation to find the beste $\lambda$ over a grid.

```{r}
yhat.boost=predict(boost.boston,newdata=Boston[-train,],n.trees=5000)
mean((yhat.boost-boston.test)^2)
boost.boston=gbm(medv~.,data=Boston[train,],distribution="gaussian",
                 n.trees=5000,interaction.depth=4,shrinkage=0.2,verbose=F)
yhat.boost=predict(boost.boston,newdata=Boston[-train,],n.trees=5000)
mean((yhat.boost-boston.test)^2)
```

---

```{r}
plot(yhat.boost,boston.test,pch=20)
abline(0,1)
mean((yhat.boost-boston.test)^2)

```

---

# Summing up

with a Kahoot quiz or another online tool (SRS?)

---




# <a id="Rpackages">R packages</a>
These packages needs to be install before knitting this R Markdown file.

```{r, eval=FALSE,echo=TRUE}
install.packages("gamlss.data")
install.packages("tidyverse")
install.packages("GGally")
install.packages("Matrix")
install.packages("tree")
install.packages("randomForest")
install.packages("gbm")
install.packages("caret")
install.packages("e1071")
```
---
 
# <a id="further">References and further reading</a>

* [Videoes on YouTube by the authors of ISL, Chapter 8](https://www.youtube.com/playlist?list=PL5-da3qGB5IB23TLuA8ZgVGC8hV8ZAdGh), and corresponding [slides](https://lagunita.stanford.edu/c4x/HumanitiesScience/StatLearning/asset/trees.pdf)
* [Solutions to exercises in the book, chapter 8](https://rstudio-pubs-static.s3.amazonaws.com/65564_925dfde884e14ef9b5735eddd16c263e.html)


---

# Acknowledgements

I thank Thea Roksv{\aa}g, whose developed the set of slides that I modified to get this set of slides. In addition, Mette Langaas and Julia Debik have contributed to the original version.

