\documentclass[english,9pt,aspectraio=169]{beamer}
\usepackage{etex}
\usetheme{uzhneu-en-informal}
%\usepackage{uarial}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\RequirePackage{graphicx,ae}
\usepackage{bm}
\usepackage{fancybox,amssymb,color}
\usepackage{pgfpages}
\usepackage{booktabs}
\usepackage{verbatim}
\usepackage{animate}
\usepackage{numprint}
\usepackage{vwcol} 
\usepackage{dsfont}
\usepackage{tikz}
\usepackage{amsmath,natbib}
\usepackage{mathbbol}
\usepackage{babel}
\usepackage{multicol}
\usepackage{xcolor}


 

\usetheme{uzhneu-en-informal}
\DeclareMathOperator{\po}{Poisson}
\DeclareMathOperator{\G}{Gamma}
\DeclareMathOperator{\Be}{Beta}
\DeclareMathOperator{\logit}{logit}
\def\n{\mathop{\mathcal N}}

\definecolor{Gray}{RGB}{139,137,137}
\definecolor{darkred}{rgb}{0.8,0,0}
\definecolor{Green}{rgb}{0,0.8,0.3}
\definecolor{lightgreen}{rgb}{0,0.7,0.3}
\definecolor{Blue}{rgb}{0,0,1}
\def\myalert{\textcolor{darkred}}
\def\myref{\textcolor{Gray}}
\setbeamercovered{invisible}

\renewcommand{\baselinestretch}{1.2}
\beamertemplateballitem
\DeclareMathOperator{\cn}{cn} % Copy number
\DeclareMathOperator{\ccn}{ccn} % common copy number
\DeclareMathOperator{\p}{p} % common copy number
\DeclareMathOperator{\E}{E} % common copy number
\DeclareMathOperator{\given}{|} % common copy number
\def\given{\,|\,}
\def\na{\tt{NA}}
\def\nin{\noindent}
\pdfpageattr{/Group <</S /Transparency /I true /CS /DeviceRGB>>}
\def\eps{\varepsilon}

\renewcommand{\P}{\operatorname{\mathsf{Pr}}} % Wahrscheinlichkeitsma√ü
\def\eps{\varepsilon}
\def\logit{\text{logit}}
%\newcommand{\E}{\mathsf{E}} % Erwartungswert
\newcommand{\Var}{\text{Var}} % Varianz
\newcommand{\Cov}{\text{Cov}} % Varianz
\newcommand{\NBin}{\text{NBin}}
\newcommand{\Po}{\text{Po}}
\newcommand{\N}{\mathsf{N}}

\newcommand{\hl}{\textcolor{Blue}}

\newcommand{\ball}[1]{\begin{pgfpicture}{-1ex}{-0.65ex}{1ex}{1ex}
\usebeamercolor[fg]{item projected}

{\pgftransformscale{1.75}\pgftext{\normalsize\pgfuseshading{bigsphere}}}
{\pgftransformshift{\pgfpoint{0pt}{0.5pt}}
\pgftext{\usebeamerfont*{item projected}{#1}}}
\end{pgfpicture}}%
\usepackage{multicol}
\newcommand{\ballsmall}[1]{\begin{pgfpicture}{-1ex}{-0.65ex}{.2ex}{.2ex}

{\pgftransformscale{1}\pgftext{\normalsize\pgfuseshading{bigsphere}}}
{\pgftransformshift{\pgfpoint{0pt}{0.5pt}}
\pgftext{\usebeamerfont*{item projected}{#1}}}
\end{pgfpicture}}%




\begin{document}
%\SweaveOpts{concordance=TRUE}

\fboxsep5pt
<<setup, include=FALSE, cache=FALSE, results='hide'>>=
library(knitr)
## set global chunk options
opts_chunk$set(fig.path='figure/', 
cache.path='cache/', fig.align='center', 
fig.show='hold', par=TRUE, fig.align='center', cache=FALSE, 
message=FALSE, 
warning=FALSE,
echo=FALSE, out.width="0.4\\linewidth", fig.width=6, fig.height=4.5, size="scriptsize", width=40)
opts_chunk$set(purl = TRUE)
knit_hooks$set(purl = hook_purl)
options(size="scriptsize")
opts_chunk$set(message = FALSE)
@

\frame{
\vspace{5mm}
\includegraphics[width=4cm]{logo_ntnu.pdf}
\vspace{-10mm}
\title[]{ \centering \huge Statistical methods in quantitative genetics}%\\[.3cm]
\author[Stefanie Muff ]{\centering Stefanie Muff}
%\institute[]{Institute of Social and Preventive Medicine \\ Institute of Evolutionary Biology and Environmental Studies}
\date[]{CAS Oslo\\ 17. March 2020}
\maketitle
}


%\frame{\tableofcontents}

\frame{\frametitle{Overview}

\begin{enumerate}
\item Genetic group models

\begin{itemize}
\item Differences in means
\item Differences in additive genetic variance (VA)
\end{itemize}

\item Marker-based regression using genomic data
\begin{itemize}
\item  
\item 
\end{itemize}

\end{enumerate}
}

\frame{\frametitle{How I became an ecological statistician}

$\rightarrow$ Master in Mathematics \\
$\rightarrow$ PhD in computational structural biology \\
$\rightarrow$ Certificate of Advanced Studies in Applied Statistics \\ 
$\rightarrow$ Postdoc in Medical and Ecological Statistics \\[2mm]

$\rightarrow$ Since Sept.\ 2019: Associate Professor in Statistics, CBD, Department of Mathematical Sciences, NTNU Trondheim \\[4mm]


I'm often not sure if I am\\[2mm]

\begin{itemize}
\item a mathematician \\[2mm]
\item a biostatistician \\[2mm]
\item an ecological statistician \\[2mm]
\item an applied statistician?\\[10mm]
\end{itemize}

}


\frame{\frametitle{Research interests}

\begin{itemize}
\item \emph{Measurement error modeling} (methods \& applications) see \emph{e.g.} \citet{muff.etal2015,muff.keller2015,muff.etal2018}.\\[4mm]
\item \emph{Bayesian statistics} (ideal for taming measurement errors!).\\[4mm]
\item \emph{Quantitative genetics} \citep{ponzi.etal2018, ponzi.etal2019,muff.etal2019}. \\[4mm]
\item \emph{Movement ecology}, see \citet{weinberger.etal2016}, \citet{gehr.etal2016} or \citet{muff.etal2019b}.\\[4mm]
\end{itemize}
}

 
\section{Genetic group animal models}

\frame{
\vspace{10mm}
\centering \LARGE \hl{Part I: Genetic group animal models}
}

\frame{\frametitle{The basic animal model}

\begin{itemize}
\item Given phenotypic measurements $y_i$ for individuals $1\leq i \leq n$, the most simple form of the animal model is 

\begin{equation*}
y_i = \mu + a_i + e_i \ ,
\end{equation*}

where $e_i \sim \N(0,\sigma_E^2)$ and $\bm{a}^\top = (a_1, \ldots, a_n)^\top \sim \N(\bm{0},\sigma_A^2 \mathbf{A})$ with additive genetic variance $\sigma_A^2$ and additive genetic relatedness matrix $\mathbf{A}$.\\[2mm]

\item The model can be extended by additional fixed or random effects.\\[2mm]

\item {\bf Assumptions}: 
 \begin{itemize}
 \item All individuals derive from the same genetic population.
 \item The breeding values ($a_i$) encode for the deviation from the mean of this population and thus have an expected value $\E[a_i]=0$.
 \end{itemize}

\end{itemize}

}


\frame{\frametitle{Systematic deviation from the assumptions}

For example

\begin{itemize}
\item in cross-bred livestock.
\item when genetically different wild subpopulations mix.
\end{itemize}



}


\frame{\frametitle{Genetic group models}

Great overview by \citet{wolak.reid2017}: \\[2mm]

\includegraphics[width=10cm]{pictures/wolak_reid.png}\\[7mm]

\myalert{Idea:} Allow for ``founder populations'' that differ in the {\bf mean} breeding value\footnote{The breeding value is the part of a phenotypic trait that has a genetic origin.}.
}




\frame{\frametitle{Example for two isolated groups}
Simple model for a phenotypic trait $y_i$ with mean $\mu$, breeding values $a_i$ and environmental component $e_i$:
%
\begin{align*} 
&\text{group 1:} \hspace{-14mm}  & y_i &= \mu +    \underbrace{a_i}_{u_i} +  e_i \ ,\\
&\text{group 2:}  \hspace{-14mm} & y_i &= \mu +  \underbrace{g_2 + a_i}_{u_i} +   e_i \ , 
\end{align*}
%
with 
%
\begin{align*} 
& u_i \sim \N(0,\sigma_A^2 \mathbf{A}) &  \hspace{-20mm}   \text{in group 1}. \\ 
& u_i \sim \N(g_2,\sigma_A^2 \mathbf{A})&  \hspace{-20mm}  \text{in group 2}. \\
\end{align*}

{\bf Interpretation:} The mean breeding values (thus probably also the mean phenotypic values) differ between the two groups.\\[2mm]

{\bf Example:} Systematic differences in wing length, weight,...

}



\frame{\frametitle{}
However, in reality there is migration between these groups.

\begin{center}
\includegraphics[width=8cm]{pictures/2populations.jpg}
\end{center}

$\Rightarrow$ Individuals have a genetic origin that stems \alert{partially from both populations}.
}


\frame{\frametitle{}
General formulation of genetic group model with $r$ groups
\begin{equation*} 
  y_i  = \mu + \sum_{j=1}^r q_{ij}g_j +  a_i + e_i  \ ,
\end{equation*}

where $q_{ij}$ is the \alert{contribution of group $j$ to the genome of individual $i$}.\\[12mm]

{\bf Caveat:} These models so far did not allow for different additive genetic variances between the populations!
}



\frame{\frametitle{Genetic group models for heterogeneous variances}
{\bf Idea}: Replace 
%
\begin{align*}
&& y_i &= \mu + \sum_{j=1}^r q_{ij}g_j + a_i  +  e_i \\[2mm]
\text{by} &\\[2mm]
&& y_i &= \mu + \sum_{j=1}^r q_{ij}g_j + \sum_{j=1}^r  q_{ij}a_{ij}  +  e_i\ ,\
\end{align*}

where $(a_{1j}, \ldots, a_{nj})^\top \sim \N(0, \sigma_{A_j}^2 \mathbf{A})$.\\[5mm]

First simulations and analyses with ingegrated nested Laplace approximations (INLA, \citet{rue.etal2009}) and MCMCglmm \citep{hadfield2010} are promising.

}


\frame{\frametitle{Inner vs outer islands}
\vspace{-4mm}
\includegraphics[angle=-90,width=10cm]{pictures/islands.pdf}\\[2mm]
}

\frame{
\vspace{10mm}
\centering \LARGE \hl{Part II: Marker-based estimation of VA using dimension-reduction techniques}
}


\frame{\frametitle{Recent developments in the animal breeding literature}
\citet{odegaard.etal2018}
}


\frame{\frametitle{Challenges in wild study systems}

In contrast to heavily selected breeding systems, wild study systems tend to have
\begin{itemize}
\item larger effective population sizes, 
\item more variable selection, 
\item higher spatio-temporal environmental and social variation.
\end{itemize}
}

\frame{\frametitle{Very short preamble on measurement error in regression models}
	~\\
	Find regression parameters $\beta_0$ and $\beta_x$ for the model with covariate $\bm{x}$:
	\begin{eqnarray*}
	y_i= \beta_0 + \beta_x \cdot x_i + \epsilon_i, \quad \epsilon_i \sim \N(0,\sigma^2_{\epsilon}) \ .
	\end{eqnarray*}
	% 	but assume that only an erroneous proxy $\bm{w}$ is observed with
	% \begin{eqnarray*}
	% w_i = x_i + u_i \qquad u_i \sim \N(0,\sigma_u^2) \quad \text{with } \quad \sigma^2_u = \sigma^2_x \ .
	% \end{eqnarray*}

	\vspace{1mm}
	\begin{center}
<<echo=F,fig.width=5, fig.height=4,out.width="9cm">>=
library(ggplot2)
set.seed(84522)
col1 <- "red"
col2 <- "blue"
n <- 100
beta_0 <- 0
beta_1 <- 1
epsilon <- rnorm(n, 0, 0.2)
x <- rnorm(n, 0, 1)
u <- rnorm(n, 0, 1)
w <- x + u
##Classical
y <- beta_0 + beta_1*x + epsilon
m1 <- lm(y~x)
m2 <- lm(y~w)
#nf <- layout(matrix(c(1,2),1,2,byrow=TRUE), c(4,4), c(4,2), TRUE)
#par( mar=c(4,6,4,1), cex.lab=1.5, cex.axis=1.4, cex.main=1.8)
cols <- c("X"="black","W"="red")
ggplot(mapping=aes(x=x,y=y,col="X")) +  geom_smooth(method="lm") + geom_point(size=0.9) + 
  # geom_smooth(mapping=aes(x=w,y=y,color="W"),method="lm") +  geom_point(mapping=aes(x=w,y=y,color="W"),size=0.9) + 
   scale_colour_manual(name="Variable",values=cols) +
   xlab("X") +
  xlim(c(-3.5,3.5)) +
  ylim(c(-2.7,2.7)) +
  theme_bw()
@
	\end{center}
}



\frame{\frametitle{Very short preamble on measurement error in regression models II}
	~\\

		However, assume that only an erroneous proxy $\bm{w}$ is observed with
	\begin{eqnarray*}
	w_i = x_i + u_i \qquad u_i \sim \N(0,\sigma_u^2) \quad \text{with } \quad \sigma^2_u = \sigma^2_x \ .
	\end{eqnarray*}

	\vspace{1mm}
	\begin{center}
<<echo=F,fig.width=5, fig.height=4,out.width="9cm">>=
library(ggplot2)
set.seed(84522)
col1 <- "red"
col2 <- "blue"
n <- 100
beta_0 <- 0
beta_1 <- 1
epsilon <- rnorm(n, 0, 0.2)
x <- rnorm(n, 0, 1)
u <- rnorm(n, 0, 1)
w <- x + u
##Classical
y <- beta_0 + beta_1*x + epsilon
m1 <- lm(y~x)
m2 <- lm(y~w)
#nf <- layout(matrix(c(1,2),1,2,byrow=TRUE), c(4,4), c(4,2), TRUE)
#par( mar=c(4,6,4,1), cex.lab=1.5, cex.axis=1.4, cex.main=1.8)
cols <- c("X"="black","W"="red")
ggplot(mapping=aes(x=x,y=y,color="X")) +  geom_smooth(method="lm") + geom_point(size=0.9) + 
  geom_smooth(mapping=aes(x=w,y=y,color="W"),method="lm") +  geom_point(mapping=aes(x=w,y=y,color="W"),size=0.9) + 
  scale_colour_manual(name="Variable",values=cols) +
  xlab("X or W") +
  theme_bw()
@
	\end{center}
}


\frame{\frametitle{Simulations and apps}

% {\bf Bias the regression parameters}, mainly attenuation (underestimation) of the true effect.\\[2mm]

Illustration with shiny apps for two error types in linear, logistic and Poisson regression:\\[6mm]

\href{https://stefaniemuff.shinyapps.io/MEC_ChooseL/}
{\beamergotobutton{Classical error}}

\href{https://stefaniemuff.shinyapps.io/MEB_ChooseL/}
{\beamergotobutton{Berkson error}}

}


\frame{\frametitle{Two running projects}
\vspace{-5mm}
{\small(With Erica Ponzi and Lukas Keller, University of Zurich)}\\[4mm]

Addressing two very different ``error problems'' in evolutionary biology:\\[4mm]

\colorbox{lightgray}{\begin{minipage}{10cm}
\begin{enumerate}
\item \myalert{Misassigned paternity error} in pedigrees

\begin{itemize}
\item has effects on estimates of {\bf inbreeding depression}, {\bf additive genetic variance} or {\bf heritability}.\\
\item Intuitive method to obtain bias-correction. \\[2mm]
\end{itemize}
$\Rightarrow$ {\bf Pedigree-SIMEX} method.\\[4mm]


\item The effect of \myalert{phenotypic measurement error} on the estimation of \\
\begin{itemize}
 \item {\bf heritability} 
 \item {\bf response to selection}
 \end{itemize}
 and how to correct for it.\\[4mm]
\end{enumerate}
\end{minipage}}
}


\frame{\frametitle{Pedigree error: the simulation extrapolation
technique meets quantitative genetics}

\begin{itemize}
\item Apparent (assumed) paternities vs.\ actual (true) paternities.\\[2mm]
\item It is known that pedigree error may lead to bias in derived quantities.\\[2mm]
\item Study system: Song Sparrows, investigated since 1975 on Mandarte Island (6ha size). Genotyping since 1993. 17\% misassigned fathers in the social pedigree \citep{reid.etal2014}.\\[8mm]
\end{itemize}

\begin{center}
\includegraphics[width=4cm]{graphics/songsparrow.jpg}\\
{\tiny \url{http://www.vogelwarte.ch/de/voegel/voegel-der-schweiz/singammer}}
\end{center}
}


\frame{\frametitle{SIMEX: A very intuitive idea}
Suggested by \citet{cook.stefanski1994}.\\[8mm]

Idea: \\[2mm]
\begin{itemize}
\item {\bf Simulation phase:} The error in the data is \alert{progressively aggravated} in order to determine how the quantity of interest is affected by the error.\\[4mm]
\item {\bf Extrapolation phase:} The observed trend is then \alert{extrapolated back} to a hypothetical error-free value.\\[10mm]
\end{itemize}

\pause
Quizz question: why is the error aggravated instead of removed??
%Mention SIMEX idea, perhaps illustrate with a covariance between two mismeasured variables.
}


\frame[containsverbatim]{\frametitle{Illustration of the SIMEX idea}

Parameter of interest: $\beta_x$ (\emph{e.g.} a regression slope).\\
Problem: The respective covariate $x$ was estimated with error: $$w=x+u\ , \quad u\sim \N(0,\sigma_u^2) \ .$$

\begin{center}
<<echo=F,fig.width=5, fig.height=4,out.width="9cm">>=
set.seed(212356)
sigmax <- 1
sigmau <- 0.25

# number of measurements
n <- 4

xx <- seq(1,3,0.25)/4#c(1,4/3,4/2,4/1) / 4
xx0 <- seq(0,3,0.05)
yy <- sigmax/(sigmax + xx)+rnorm(length(xx),0,0.03)
yy0 <- sigmax/(sigmax + xx0)

par(mar=c(4,5,2,1))
plot(4*xx,yy,ylim=c(0.4,1.1),xlim=c(0,3.25),xlab=expression(sigma[u]^2),ylab=expression(beta[x]),cex.lab=1.24,cex=1.2)
#axis(1,tick=T,at=c(0,xx*4))#,labels=c(0,1,1.33,2,4))
points(4*xx[1],yy[1],pch=18,cex=2.5,col=2)
points(0,1,pch=18,cex=2.5,col=3)
legend("topright",legend=c("Naive","Corrected"),pch=18,col=c(2,3),cex=1.2)
lines(4*xx0,yy0,lty=2,lwd=2)
abline(v=1,lty=3,lwd=1)
@
\end{center}
}


\frame{\frametitle{Pedigree-SIMEX}

The idea can be transferred to pedigree error by a \myalert{successive increase of the error proportion} and \myalert{extrapolation} to zero error. 
\vspace{-2mm}

\begin{center}
\includegraphics[width=6.9cm,height=6.1cm]{graphics/sparrows2.pdf}
\end{center}
\vspace{-0.5cm}
$\rightarrow$ \texttt{PSIMEX} package on CRAN (written by Erica Ponzi).
}


\frame{\frametitle{P-SIMEX without a pedigree?}
Thanks to the availability of SNP data, pedigrees may be replaced by \myalert{genomic relatedness matrices}. \\[4mm]

Are these estimates of relatedness expected to be error-free? \\[4mm]

\pause
$\rightarrow$ Sampling error, imperfect tagging, ... \\[6mm]

\begin{center}
\colorbox{lightgray}{\begin{minipage}{10cm}
Idea: Use the SIMEX-approach on the entries of the relatedness matrix....
\end{minipage}}
\end{center}
}





\frame{\frametitle{ }
\begin{center}
{\Large THANK YOU!}
\end{center}
}



\frame[allowframebreaks]{References:\\[2mm]
{\scriptsize
\bibliographystyle{Chicago}
\bibliography{refs}
}
}

\end{document}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

